content/subtleties.md

```diff
@@ -14,37 +14,39 @@ Also see Pierre Weis's [Frequently asked Questions about Caml](http://caml.inria
   What is the difference between a function with (for example) type `'a -> 'b`,
   and one with type `'_weak1 -> '_weak2`?  Why am I getting a
   compiler error like these? (It worked when I tried it out in utop!)
   ```
-  Values do not match:
-    val foo : int -> int
-  is not included in
-    val foo : 'a -> 'a
-  
   Values do not match:
     val foo : '_weak1 -> '_weak1
   is not included in
     val foo : 'a -> 'a
   ```
 
-  Part of the answer is that `'_weak1`, `'_weak2`, etc., which are called
-  "weak types" or "weakly polymorphic types", are roughly speaking, only *temporarily*
-  [polymorphic](http://ocaml.org/learn/tutorials/basics.html#Polymorphic-functions) types (unlike `'a`, etc.).  If a function's type includes a weak type,
-  the compiler will permanently replace the weak type with a regular concrete type
-  (e.g. `float list`) the first time that it sees an application of the function.
-  The errors above say that the type `'a -> 'a` specified by a signature is too general.
-  The signature says that the function will work with any type, but the compiler
-  thinks that your function might only work with a single input and output type
-  (`int`, in the first error).  The value restriction is the name of the heuristic
-  that the compiler uses to make this decision.  For more information see the section on
-  [Side effects and weak polymorphism](https://realworldocaml.org/v1/en/html/imperative-programming-1.html#side-effects-and-weak-polymorphism)
-  in *Real World OCaml*,
-  [Section 5.1 Weak polymorphism and mutation](http://caml.inria.fr/pub/docs/manual-ocaml/polymorphism.html#sec51) 
-  in the OCaml Manual, [Common Error Messages](http://ocaml.org/learn/tutorials/common_error_messages.html#The-type-of-this-expression-contains-type-variables-that-cannot-be-generalized) at ocaml.org, and ["My function is not polymorphic (enough) ?"](http://caml.inria.fr/pub/old_caml_site/FAQ/FAQ_EXPERT-eng.html#eta_expansion) and subsequent sections in "Frequently asked Questions about Caml".
+  Part of the answer is that `'_weak1`, `'_weak2`, etc., which are called "weak type variables", are roughly speaking, only *temporarily* [polymorphic](http://ocaml.org/learn/tutorials/basics.html#Polymorphic-functions) types (unlike `'a`, etc.).  If an expression type includes a weak type variable, the compiler will permanently substitute the variable with a regular concrete (i.e., monorphic) type (e.g. `float list`) the first time that it sees an application of the function.
+  
+  In OCaml and other ML-style languages that support mutability, only _values_ can have polymorphic types, hence the name -- "value restriction". The underlying reason, is that if an expression is mutable (an since we don't know its type, we must assume that it is mutable), then it could be assigned multiple times, each time with a value of a different type, thus we can't guarantee that the type of a value would be preserved. That will eventually break the type soundness. 
```

Comment:
When you say "only _values_ can have polymorphic types", you should note an example of something else that cannot. Perhaps even include two code snippets, one commented "can have a polymorphic type", "cannot have a polymorphic type".

Also perhaps show an example of a mutable expression, and an attempt to assign to it with two different type values, thus showing the problem?

Reply:
will do